Transition font crate to 2018 edition

sterlingjensen opened the pull request:
In addition to implementing the path changes, dependencies are now more strictly specified, and a light linting was performed.

chrisduerr commented on font/src/darwin.rs:
```diff
@@ -40,14 +40,49 @@ use core_text::font_descriptor::SymbolicTraitAccessors;
+            .map(|(_, val)| val)
+            .collect::<Vec<_>>()
+    }
+}
```
What's the intention behind moving this into the default module? Usually modules in the same file are not desired except for test modules.

sterlingjensen replied:
My intention was to simplify the pre-2018 mod structuring (folder, ambiguous mod.rs file, meaningfully named submodules), while leaving the module path structure unchanged. I can only guess at the original intent, likely a combination of crate constraints and unrealized future plans that needed the scoping. As I can't test for macos before pushing, I'm reluctant to mess around with it more than necessary.

chrisduerr replied:
> ambiguous mod.rs file

What do you mean by this? How is a `mod.rs` file ever ambiguous?

Modules should never be in the same file without a good reason for it. That's why it originally was put into a different module, since it's unrelated code that has a nice separation from the primary module.

sterlingjensen replied:
`mod.rs` files require a path in order to be distinguished from one another, this was sufficiently annoying to enough people that a new naming scheme was offered.

I'm pretty surprised about the mod aversion, it seems like the perfect application in this case - grouping a small amount of like code in the one file it is used, a grep of this crate's dependencies shows hundreds of non-test mod blocks. I'll revert the darwin directory.

chrisduerr replied:
The only reason why you would ever go without a `mod.rs` file is when you have only a single module. We have more than one module here, so there's absolutely no point in trying to cram everything into one massive file.

The `mod.rs` file still has its purpose for structures with more than a single module in one directory.

nixpulvis replied:
FWIW, I don't think `mod.rs` is ambiguous, however it is somewhat _redundant_, which is why the Rust team has added the new resolution in 2018. If I had to choose myself, I'd be in favor of the old style. The main justification is that it's consistent with `lib.rs`.

For small personal projects I can see using the new style, but I think I agree we shouldn't make this change.

chrisduerr commented on font/src/ft/fc.rs:
```diff
@@ -11,42 +11,34 @@
+    {FC_WEIGHT_BLACK, FC_WEIGHT_BOLD, FC_WEIGHT_EXTRABLACK, FC_WEIGHT_EXTRABOLD},
+    {FC_WEIGHT_BOOK, FC_WEIGHT_MEDIUM, FC_WEIGHT_REGULAR, FC_WEIGHT_SEMIBOLD},
+    {FC_WEIGHT_EXTRALIGHT, FC_WEIGHT_LIGHT, FC_WEIGHT_THIN},
+};
```
Can't we remove all brackets in here?

sterlingjensen replied:
I'd be happy to, I just thought it safer to preserve the grouping - as it seemed to me that someone wanted them grouped logically instead of alphabetically.

chrisduerr replied:
The current grouping does nothing, so there's no point in keeping it.

chrisduerr commented on font/src/ft/fc.rs:
```diff
@@ -11,42 +11,34 @@
-
+use std::{fmt, ptr};
+
+use fontconfig::fontconfig::{
```
Having a `fontconfig::fontconfig` import is very confusing, which is why the original version had `use fontconfig::fontconfig as ffi` as an easier to understand import. I think that makes a lot of sense and makes it a ton easier to understand what exactly we're importing here. So I'd be in favor of changing that back.

sterlingjensen replied:
I was actually tempted to rebind it to the name of the crate it came from. I'll add ffi back in and bind off of that.

chrisduerr commented on font/src/ft/fc/pattern.rs:
```diff
@@ -11,23 +11,23 @@
+    path::PathBuf,
+    ptr::{self, NonNull},
+    str,
+};
```
We do not use nested imports in Alacritty like this. They're too difficult to read without much benefits at all.

sterlingjensen replied:
IntelliJ must be back to its old tricks after the update, I'll correct that.

chrisduerr commented on font/src/ft/fc.rs:
```diff
@@ -11,42 +11,34 @@
+pub use config::{Config, ConfigRef};
+pub use font_set::{FontSet, FontSetRef};
+pub use object_set::{ObjectSet, ObjectSetRef};
+pub use pattern::{Pattern, PatternRef};
```
This is done incorrectly in the latest master too, but generally we do not want to use `pub use` excessively. Imports like `self::Pattern` make it extremely difficult to understand where things are coming from. Meanwhile `pattern::Pattern` is without a doubt easier to understand.

sterlingjensen replied:
I've got no problem with that, just trying to preserve the existing paths. Do you want me to refactor the visibility in this merge request? Because I was already getting the feeling that I had touched too many files.

chrisduerr replied:
Feel free to do either, I'd understand not wanting to touch any existing code if the goal is just to transition to a new edition.

However when it comes to style improvements, I'm always willing to make exceptions since these should just improve codestyle in general.

chrisduerr commented on font/src/ft/fc/pattern.rs:
```diff
@@ -528,7 +528,7 @@ impl PatternRef {
     }
 
-    pub fn render_prepare(&self, config: &ConfigRef, request: &PatternRef) -> Pattern {
+    pub fn render_prepare(&self, config: &ConfigRef, request: &PatternRef) -> self::Pattern {
```
There's three instances of `self::Pattern` in this file, however the module itself is `pattern.rs`. Shouldn't the `Pattern` be defined in this file, thus not needing any prefix at all?

sterlingjensen replied:
I thought so as well, and tried to figure out what the problem was - before I gave up. Rustc is clever enough that the prefix is unneeded for compilation. but I have found RLS and IntelliJ to be much less clever about macro expansion (my guess at the root cause, the foreign_type macro). This was the case on both Windows and Debian, stable and nightly. So should I go with the syntactically correct original, or the practical existing submission?

chrisduerr replied:
We do not need to cater to RLS and IntelliJ, our code is compiled by rustc. So if the IDEs and editors don't understand it, that's their problem not ours.

sterlingjensen replied:
Ok, I'll revert the catering.

chrisduerr commented on font/src/ft/fc.rs:
```diff
@@ -11,42 +11,35 @@
 ///
 /// The returned pattern is the result of Pattern::render_prepare.
-pub fn font_match(config: &ConfigRef, pattern: &mut PatternRef) -> Option<Pattern> {
+pub fn font_match(config: &ConfigRef, pattern: &mut PatternRef) -> Option<self::Pattern> {
```
```suggestion
pub fn font_match(config: &ConfigRef, pattern: &mut PatternRef) -> Option<Pattern> {
```

chrisduerr commented on font/src/ft/fc/pattern.rs:
```diff
@@ -18,16 +18,15 @@ use std::path::PathBuf;
+    FcPatternAddCharSet, FcPatternAddDouble, FcPatternAddInteger, FcPatternAddString,
+    FcPatternCreate, FcPatternDestroy, FcPatternGetBool, FcPatternGetDouble, FcPatternGetInteger,
+    FcPatternGetString, FcPatternPrint, FcResultMatch,
+};
```
Here we use `fontconfig::fontconfig` again, instead of `super::ffi`, which is inconsistent.

Generally I think that these `fontconfig::fontconfig` imports are an insane mess. We have cross-module depedencies for the sole reason of using a crate import rename which seems to originate from the old edition, but just doesn't work well with the new 2018 edition.

I think the reasonable thing to do would be to import `fontconfig::fontconfig` everywhere. There's just no way around that and it's just imports so it shouldn't matter too much. If we ever import `fontconfig::fontconfig::self` anywhere, then we can rename that to `fontconfig::fontconfig::self as fontconfig_ffi`.

However we should never use `super::fontconfig::fontconfig` or `self::fontconfig::fontconfig` anywhere, since that just unreasonably complicates the cross-module depedencies for no reason.

Sorry for going back and forth on this, but if we want to have a consistent code style, we need to make sure it's intuitive and `super::ffi::` definitely is not that.

chrisduerr commented on font/src/ft.rs:
```diff
@@ -21,6 +21,7 @@ use std::path::PathBuf;
 use freetype::tt_os2::TrueTypeOS2Table;
```
This file needs to go back to `font/src/ft/mod.rs`.

chrisduerr commented on font/src/directwrite.rs:
```diff
@@ -13,10 +13,7 @@
 // limitations under the License.
```
This file should go back to `font/src/directwrite/mod.rs` for platform consistency.

chrisduerr commented on font/src/ft/fc.rs:
```diff
@@ -11,42 +11,35 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
```
This file needs to go back to `font/src/ft/fc/mod.rs`.

sterlingjensen commented:
With all the suggested changes, a single 2018 edition improvement remains... a solitary crate scope absolute path. It would be inaccurate to describe this as a "transition" when platform consistency means 2015 conventions.

chrisduerr commented:
> It would be inaccurate to describe this as a "transition" when platform consistency means 2015 conventions.

Only because it is supported to have both `foo.rs` and the `foo` directory, does not mean that the `foo/mod.rs` structure is deprecated.

Putting things into the same path is a far cleaner solution, regardless of 2018 edition or not and having both the directory and the module on the top level does not scale.

sterlingjensen commented:
I'd say that two birds could be killed with one stone, simplified paths and cleaned up `pub use`s, but your justification for `byte_order.rs` similarly justifies all those interlinked `fc` sub-modules.

nixpulvis commented:
@sterlingjensen I'd be happy to see this finished up and merged. Would you be interested in continuing it with the changes @chrisduerr has suggested? Or should I try and complete it?

sterlingjensen commented:
I don't think I should be touching this part of the code base, as it appears I'm in the minority on some very fundamental issues :upside_down_face: Please feel free to salvage whatever you can.
